pinot-query-planner/src/test/java/org/apache/pinot/query/planner/logical/StagesTestBase.java

```diff
@@ -0,0 +1,230 @@
+   * test is getting too complex and difficult to read and/or get insights from it in case of failure.
+   */
+  @AfterMethod
+  public void setUp() {
+    _stageRoots.clear();
+  }
+
+  /**
+   * Creates a new plan that will have an initial stage.
+   *
+   * The stage will have a default {@link MailboxSendNode} whose stage will be 0 and its child the one created by the
+   * builder.
+   *
+   * Notice that this method does not offer any way to customize the initial send mailbox.
+   */
+  public MailboxSendNode when(SimpleChildBuilder<? extends PlanNode> builder) {
+    return sendMailbox(0, builder).build(0);
+  }
+
+  /**
+   * Returns a builder that can be used to create a new mailbox receive node.
+   *
+   * It is usually recommended to use {@link #exchange(int, SimpleChildBuilder)} instead of this method, given that
+   * {@code exchange} creates a pair of send and receive mailboxes and deals with the stageId management.
+   */
+  public SimpleChildBuilder<MailboxReceiveNode> receiveMailbox(SimpleChildBuilder<MailboxSendNode> childBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      MailboxSendNode mailbox = childBuilder.build(stageId);
+      int nextStageId = mailbox.getStageId();
+      return new MailboxReceiveNode(stageId, mySchema, List.of(), nextStageId, null, null, null, null, false, false,
+          mailbox);
+    };
+  }
+
+  /**
+   * Creates a join node that will have the left and right nodes as children.
+   *
+   * The join type will be {@link JoinRelType#FULL}, the join strategy will be {@link JoinNode.JoinStrategy#HASH} and
+   * there will be no conditions. If custom joins are needed feel free to add more builder methods or create your own
+   * instance of {@link SimpleChildBuilder}.
+   */
+  public SimpleChildBuilder<JoinNode> join(
+      SimpleChildBuilder<? extends PlanNode> leftBuilder,
+      SimpleChildBuilder<? extends PlanNode> rightBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      PlanNode left = leftBuilder.build(stageId);
+      PlanNode right = rightBuilder.build(stageId);
+      return new JoinNode(stageId, mySchema, myHints, List.of(left, right), JoinRelType.FULL,
+          Collections.emptyList(), Collections.emptyList(), Collections.emptyList(), JoinNode.JoinStrategy.HASH);
+    };
+  }
+
+  /**
+   * Creates a pair of receiver and sender nodes that will be logically connected.
+   *
+   * Whenever this builder is used to create a node, the mailbox send node will be added to the list of mailboxes.
+   *
+   * Although there are builder methods to create send and receive mailboxes separately, this method is recommended
+   * because it deals with the stageId management and creates tests that are easier to read.
+   */
+  public SimpleChildBuilder<MailboxReceiveNode> exchange(
+      int nextStageId, SimpleChildBuilder<? extends PlanNode> childBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      PlanNode input = childBuilder.build(stageId);
```

Comment:
I think this should be `childBuilder.build(nextStageId)` right?

The child is the input to the `MailboxSendNode` and must be in the same stage as it. For instance, in `EquivalentStagesFinderTest`, a test like 

```
when(
        join(
            exchange(1, tableScan("T1")),
            exchange(2, tableScan("T2"))
        )
    );
```

results in the join node as well as both the table scan nodes being in stage 0, but the table scan nodes should be in stages 1 and 2.

Reply:
You are right. Fixing it